addons/purchase/tests/test_purchase.py

```diff
@@ -336,3 +336,27 @@ def test_purchase_not_creating_useless_product_vendor(self):
         po.button_confirm()
 
         self.assertEqual(po.order_line.product_id.seller_ids.mapped('name'), delivery_address)
+
+    def test_product_purchase_method_with_receive_as_default_purchase_method(self):
+        self.env['ir.default'].set('product.template', 'purchase_method', 'receive', company_id=True)
+
+        product = self.env['product.product'].create({'name': 'product_test'})
+        self.assertEqual(product.purchase_method, 'receive')
+
+        product.update({'detailed_type': 'service'})
+        self.assertEqual(product.purchase_method, 'purchase')
+
+        product.update({'detailed_type': 'product'})
+        self.assertEqual(product.purchase_method, 'receive')
+
+    def test_product_purchase_method_with_purchase_as_default_purchase_method(self):
+        self.env['ir.default'].set('product.template', 'purchase_method', 'purchase', company_id=True)
+
+        product = self.env['product.product'].create({'name': 'product_test'})
+        self.assertEqual(product.purchase_method, 'purchase')
+
+        product.update({'detailed_type': 'service'})
+        self.assertEqual(product.purchase_method, 'purchase')
+
+        product.update({'detailed_type': 'product'})
+        self.assertEqual(product.purchase_method, 'purchase')
```

Comment:
Hey @AH-Yussef it looks like you made a little mistake here. `'detailed_type': 'product'` is added in stock so you need to put these tests in `purchase_stock`.

Also, there's no reason to use `.update()` in these tests, use `write` instead.